std/gkrapi/compile.go

```diff
@@ -5,236 +5,282 @@ import (
 	"fmt"
 	"math/bits"
 
-	"github.com/consensys/gnark/constraint/solver"
+	"github.com/consensys/gnark-crypto/ecc"
 	"github.com/consensys/gnark/constraint/solver/gkrgates"
 	"github.com/consensys/gnark/frontend"
 	gadget "github.com/consensys/gnark/internal/gkr"
+	"github.com/consensys/gnark/internal/gkr/gkrhints"
 	"github.com/consensys/gnark/internal/gkr/gkrinfo"
 	"github.com/consensys/gnark/internal/gkr/gkrtypes"
 	"github.com/consensys/gnark/internal/utils"
 	fiatshamir "github.com/consensys/gnark/std/fiat-shamir"
 	"github.com/consensys/gnark/std/gkrapi/gkr"
 	"github.com/consensys/gnark/std/hash"
+	"github.com/consensys/gnark/std/multicommit"
 )
 
 type circuitDataForSnark struct {
 	circuit     gkrtypes.Circuit
 	assignments gkrtypes.WireAssignment
 }
 
-type Solution struct {
-	toStore      gkrinfo.StoringInfo
-	assignments  gkrtypes.WireAssignment
-	parentApi    frontend.API
-	permutations gkrinfo.Permutations
+type InitialChallengeGetter func() []frontend.Variable
+
+// Circuit represents a GKR circuit.
+type Circuit struct {
+	toStore              *gkrinfo.StoringInfo
+	assignments          gkrtypes.WireAssignment
+	getInitialChallenges InitialChallengeGetter // optional getter for the initial Fiat-Shamir challenge
+	ins                  []gkr.Variable
+	outs                 []gkr.Variable
+	api                  frontend.API              // the parent API used for hints
+	hints                *gkrhints.TestEngineHints // hints for the GKR circuit, used for testing purposes
+	index                int                       // index among all GKR circuits
 }
 
-func (api *API) nbInstances() int {
-	if len(api.assignments) == 0 {
-		return -1
+// New creates a new GKR API
+func New(api frontend.API) *API {
+	toStore, index := api.(gkrinfo.ConstraintSystem).NewGkr()
```

Comment:
I would handle the error better - right now if the API doesn't implement it we just get "doesn't implement the interface" panic, but it doesn't allow user to fix the issue (perhaps they wrapped their builder for adding functionality etc)